### PR TITLE
Hotfix: limit PDF render threads to 1. Prevents frequent crashes for some users.

### PR DESCRIPTION
**Pull request by Myridium**
HOTFIX 
Git diff contributed by @astoeckel in #1409 . Thank you for your contribution.

This commit limits the render threads to 1 in PDF generation. This prevents frequent crashing for users with high core counts, where the (undiagnosed; suspected) threading issue appears to be worse. 

For me, `texstudio` was crashing ~30% of the time the PDF was generated. This commit has eliminated the crashing completely. **I do not notice any difference in the render time**, so I wonder whether the multithreaded PDF generation is needed or working anyway. I haven't looked into it. All I know is that a working `texstudio` is better than one that crashes every third compile.

### CODE REVIEW

**Comment by sunderme** [comment]
as the vast majority of users does not see that crash, I don‘t see the point in limiting performance for everyone.
So, a setting in texstudio.ini should be used and communicated. I thought that there already is one, but I am not sure.

**Comment by Myridium** [comment]
@sunderme like I said I notice no performance difference. I am typesetting a 100 page document with several images. I am not sure if the multithreading was even worknig properly to begin with.

Obviously it is a bug and should be fixed. I'm disappointed that you can't see that.

**Comment by Myridium** [comment]
And you should expect this to start affecting more users. I doubt you know how many users it is affecting. Most users who try a program that crashes will move on to another. With the latest generation of AMD processors we can expect this to come up much more frequently.

**Comment by astoeckel** [comment]
Two thoughts. First, I think that only the changes to `src/pdfviewer/PDFDocument.cpp` are required, the patch limiting the global Qt thread pool size has no effect.

Second, I agree with Myridium's assessment that a) the perceived performance impact of single thread PDF rendering is negligible (and I have used TeX Studio with a single rendering thread on large documents for more than a month now) and b) that users having the same issue are unlikely to report this on GitHub; they may just move to another editor. I was also moving to another editor for a while after TeX Studio started crashing frequently, and only came back because TeX Studio crashing every ten minutes was still better than the other editors.

To summarize, we currently have the following situation:

Low thread count CPU -> Intrinsically small improvement in PDF rendering performance due to multi threading
High thread count CPU -> Potentially faster PDF rendering, but frequent crashes due to multi threading not being implemented correctly

I'd propose limiting the thread count to one per default, and telling users about an ini setting to re-enable multithreaded rendering at their own risk.

**Comment by Myridium** [comment]
Thanks @astoeckel, I am currently recompiling texstudio with that extraneous line removed. I'll test it a bit then add the commit.

**Comment by sunderme** [comment]
use the latest (dev) version of txs
There is a texstudio.ini only configuration which can limit the number of pdf threads to one. `Preview/LimitRenderQueues=1`
default is min(number of cores,8)

**Comment by Myridium** [comment]
So rather than fix a known bug, or even hotfix it, you're going to tell users to implement a hotfix themself? What?

**Comment by sunderme** [comment]
You are aware that it crashes in poppler which is (usually) thread-safe and the "fix" is basically a work-around which disables multi-threaded access to said library ?

**Comment by astoeckel** [comment]
First of all, thank you @sunderme  for looking into this!

Poppler is only partially thread safe. For example, there can only be one thread accessing a single Page object at a time, see [1]. My gut feeling is that there is a race condition in the TeXstudio rendering code that, under rare circumstances, let's different threads access the same page at the same time. In other words, I wouldn't blame upstream for this before this issue is not more thoroughly investigated. I posted some code in #1409 that allows you to trigger the bug quite reproducibly, even within a debugging environment.

Limiting the number of TeXstudio render threads to eight per default may be a reasonable compromise, but I still don't feel good about this.

[1] https://gitlab.freedesktop.org/poppler/poppler/-/issues/666#note_76932

**Comment by sunderme** [comment]
each render engine gets its own page, renders it and then deletes it. The code is in use for several years.
I don't see crashes with 16 threads and crashes are usually reported.
So, something is special why it crashes on your side. Unless that is cleared up, there is not much I can do.
I **have** provided your work-around in a more general way, but I don't see the necessity to activate it for everyone when they don't need it.

**Comment by astoeckel** [comment]
Just to be sure, you are also not able to reproduce the bug when you disable the page cache as I described in #1409?

**Comment by sunderme** [comment]
okay, if I disable caching and reduce size , so that several pages are visible, I can reproduce the crash.
It is still a poppler crash, apparently when parsing for font information. The debugger shows that 8 distinct pages are rendered.
Next step would be a test program which renders and crashes, though not sure if it would do so automatically.

**Comment by astoeckel** [comment]
Good to know! By test program, do you mean a headless CLI program that hammers PDFRenderEngine to evoke the crash? If you'd like to, I can take a stab at that this weekend.

**Comment by sunderme** [comment]
yes, that is what I mean.
